### PR TITLE
Remove invalid assignments after worker/shift edit

### LINKED ISSUE #109: Remove assignments that are no longer valid after commands
Remove assignments if:

- [x] The worker is unavailable for the shift after a `ShiftEditCommand` changes the `ShiftDay`/`ShiftTime`(#105)
- [x] The worker is unavailable for the shift after a `WorkerEditCommand` changes the worker's unavailabilities (#105)
- [x] The worker is not fit for the role (ie. the worker's role set does not contain the role) after a `WorkerEditCommand` changes the worker's role set
- [x] The shift does not require/require fewer of that role after a `ShiftEditCommand` changes the shift's role requirements (require user to provide a list of worker indexes to unassign from the shift when reducing the quantity needed for a role)

### PR DESCRIPTION
**Pull request by tnsyn**
Resolves #109

### CODE REVIEW

File: src/main/java/seedu/address/logic/commands/WorkerEditCommand.java [hunk 1]
```diff
@@ -130,12 +129,22 @@ private void editWorkerInAssignments(Model model, Worker workerToEdit, Worker ed
...
         List<Assignment> fullAssignmentList = model.getFullAssignmentList();
         List<Assignment> assignmentsToEdit = new ArrayList<>();
+        List<Assignment> assignmentsToDelete = new ArrayList<>();
 
         for (Assignment assignment : fullAssignmentList) {
             if (workerToEdit.isSameWorker(assignment.getWorker())) {
-                assignmentsToEdit.add(assignment);
+                Role assignedRole = assignment.getRole();
+                Set<Role> editedWorkerRoles = editedWorker.getRoles();
+                if (!editedWorkerRoles.contains(assignedRole)) {
```
Comment (wakululuu) [inline src/main/java/seedu/address/logic/commands/WorkerEditCommand.java]:
```suggestion
                if (!editedWorker.isFitForRole(assignment.getRole())) {
```

File: src/main/java/seedu/address/logic/commands/ShiftEditCommand.java [hunk 2]
```diff
@@ -107,12 +108,24 @@ private void editShiftInAssignments(Model model, Shift shiftToEdit, Shift edited
...
         List<Assignment> assignmentsToEdit = new ArrayList<>();
+        List<Assignment> assignmentsToDelete = new ArrayList<>();
+        Set<Role> newRoles = new HashSet<>();
 
         for (Assignment assignment : fullAssignmentList) {
             if (shiftToEdit.isSameShift(assignment.getShift())) {
-                assignmentsToEdit.add(assignment);
+                Role assignmentRole = assignment.getRole();
+                Set<RoleRequirement> newRoleRequirements = editedShift.getRoleRequirements();
+                newRoleRequirements.forEach(roleRequirement -> newRoles.add(roleRequirement.getRole()));
```
Comment (wakululuu) [inline src/main/java/seedu/address/logic/commands/ShiftEditCommand.java]: Perhaps these two lines can be moved out of and before the `for` loop.

File: src/main/java/seedu/address/logic/commands/ShiftEditCommand.java [hunk 3]
```diff
@@ -107,12 +108,24 @@ private void editShiftInAssignments(Model model, Shift shiftToEdit, Shift edited
...
+                if (!newRoles.contains(assignmentRole)) {
+                    assignmentsToDelete.add(assignment);
```
Comment (wakululuu) [inline src/main/java/seedu/address/logic/commands/ShiftEditCommand.java]: Seems like this only works for the case where a role requirement that was present in a shift is completely removed from it. I don't think the case where the `quantityRequired` of a role is changed but the `role` remains unchanged has been accounted for.

**Review by wakululuu — CHANGES_REQUESTED** [review]
`WorkerEditCommand` looks good but I think there is a somewhat major problem with `ShiftEditCommand`.

File: src/main/java/seedu/address/logic/commands/ShiftEditCommand.java (hunk 3 above)
Comment (tnsyn) [reply src/main/java/seedu/address/logic/commands/ShiftEditCommand.java]: Yes, thanks for reminding me! Will add the checks for this

File: src/main/java/seedu/address/logic/commands/ShiftEditCommand.java (hunk 2 above)
Comment (tnsyn) [reply src/main/java/seedu/address/logic/commands/ShiftEditCommand.java]: I've made this into a method called getRoles() inside the Shift class!

File: src/main/java/seedu/address/logic/commands/WorkerEditCommand.java (hunk 1 above)
Comment (tnsyn) [reply src/main/java/seedu/address/logic/commands/WorkerEditCommand.java]: Looks better, will change!

**Review by wakululuu — APPROVED** [review]
LGTM other than the merge conflict to be resolved

**Pull request merged**